File: api/api-javascript/api-platform/src/main/resources/META-INF/dirigible/platform/v4/template-engines.js
/*
 * Copyright (c) 2021 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
/**
 * API v4 Template Engine
 * 
 * Note: This module is supported only with the Mozilla Rhino engine
 */

var repository = require("platform/v4/repository");

const REGISTRY_PUBLIC = "/registry/public/";
const MUSTACHE_FILE_EXTENSION = ".mustache";

exports.getDefaultEngine = function() {
	return exports.getVelocityEngine();
};

exports.getMustacheEngine = function() {
	var engine = org.eclipse.dirigible.api.v3.platform.TemplateEnginesFacade.getMustacheEngine();
	return new TemplateEngine(engine, "mustache");
};

exports.getVelocityEngine = function() {
	var engine = org.eclipse.dirigible.api.v3.platform.TemplateEnginesFacade.getVelocityEngine();
	return new TemplateEngine(engine, "velocity");
};

exports.getJavascriptEngine = function() {
	var engine = org.eclipse.dirigible.api.v3.platform.TemplateEnginesFacade.getJavascriptEngine();
	return new TemplateEngine(engine, "javascript");
};

exports.generate = function(template, parameters) {
	return exports.getDefaultEngine().generate(template, parameters);
};

exports.generateFromFile = function(location, parameters) {
	var resource = repository.getResource(REGISTRY_PUBLIC + location);
	if (resource.exists()) {
		var isMustacheTemplate = location.endsWith(MUSTACHE_FILE_EXTENSION);
		var engine = isMustacheTemplate ? exports.getMustacheEngine() : exports.getDefaultEngine();
		var template = resource.getText();
		return engine.generate(template, parameters);
	}
	return null;
};

function TemplateEngine(engine, type) {
	this.engine = engine;
	this.sm = type === "mustache" ? "{{" : null;
	this.em = type === "mustache" ? "}}" : null;

	this.generate = function(template, parameters) {
		return this.engine.generate(template, JSON.stringify(parameters), this.sm, this.em);
	};

	this.setSm = function(sm) {
		this.sm = sm;
	};

	this.setEm = function(em) {
		this.em = em;
	};
}

File: components/api/api-modules-java/src/main/java/org/eclipse/dirigible/sdk/template/TemplateEngines.java
/*
 * Copyright (c) 2010-2026 Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: Eclipse Dirigible contributors SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.dirigible.sdk.template;

import java.io.IOException;
import org.eclipse.dirigible.components.api.templates.TemplateEnginesFacade;

/**
 * Renders a template against a JSON parameter document. Three engines are available:
 *
 * <ul>
 * <li><strong>Mustache</strong> — logic-less, perfect for emails and HTML fragments.</li>
 * <li><strong>Velocity</strong> — Apache Velocity with the platform's default tool set.</li>
 * <li><strong>JavaScript</strong> — evaluates {@code ${expr}} fragments through GraalJS for the
 * full power of script expressions inside the template.</li>
 * </ul>
 *
 * {@link #generate(String, String)} picks the default engine (Mustache); the named variants are for
 * explicit choice. {@link #generate(String, String, String, String, String)} lets you change the
 * marker pair (defaults are <code>&#123;&#123; &#125;&#125;</code> for Mustache) — useful when the
 * template body itself contains delimiter characters.
 */
public final class TemplateEngines {

    private TemplateEngines() {}

    public static String generate(String template, String parametersJson) throws IOException {
        return TemplateEnginesFacade.getDefaultEngine()
                                    .generate(null, template, parametersJson);
    }

    public static String generateMustache(String template, String parametersJson) throws IOException {
        return TemplateEnginesFacade.getMustacheEngine()
                                    .generate(null, template, parametersJson);
    }

    public static String generateVelocity(String template, String parametersJson) throws IOException {
        return TemplateEnginesFacade.getVelocityEngine()
                                    .generate(null, template, parametersJson);
    }

    public static String generateJavascript(String template, String parametersJson) throws IOException {
        return TemplateEnginesFacade.getJavascriptEngine()
                                    .generate(null, template, parametersJson);
    }

    public static String generate(String location, String template, String parametersJson, String startMarker, String endMarker)
            throws IOException {
        return TemplateEnginesFacade.getDefaultEngine()
                                    .generate(location, template, parametersJson, startMarker, endMarker);
    }
}
